File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.cpp
class Solution {
public:
    int countUnguarded(int m, int n, vector<vector<int>>& guards, vector<vector<int>>& walls) {
        int g[m][n];
        memset(g, 0, sizeof(g));

        for(auto& e : guards){
            g[e[0]][e[1]] = 2;
        }
        for(auto& e : walls){
            g[e[0]][e[1]] = 2;
        }

        int dirs[5] = {-1,0,1,0,-1};
        for(auto& e: guards){
            for(int k=0; k<4; ++k){
                int x = e[0], y=e[1];
                int dx = dirs[k], dy= dirs[k+1];

                while(x+dx >= 0 && x+ dx < m && y+dy >=0 && y+dy <n && g[x+dx][y+dy] <2){
                    x += dx;
                    y += dy;
                    g[x][y] = 1;
                }
            }
        }

        int unguardedCount =0;
        for(int i=0; i<m; i++){
            unguardedCount += count(g[i], g[i]+n, 0);
        }

        return unguardedCount;
    }
};

File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.java
import java.util.*;

class Solution {
    void dfs(int r, int c, String dir, int[][] vis, Map<String, Integer> mp) {
        int n = vis.length;
        int m = vis[0].length;
        if (r < 0 || c < 0 || r >= n || c >= m) return;
        if (mp.containsKey(r + "," + c)) return;
        vis[r][c] = 1;

        if (dir.equals("r")) dfs(r, c + 1, "r", vis, mp);
        if (dir.equals("l")) dfs(r, c - 1, "l", vis, mp);
        if (dir.equals("u")) dfs(r - 1, c, "u", vis, mp);
        if (dir.equals("d")) dfs(r + 1, c, "d", vis, mp);
    }

    public int countUnguarded(int m, int n, int[][] guards, int[][] walls) {
        int[][] vis = new int[m][n];
        Queue<int[]> q = new LinkedList<>();
        Map<String, Integer> mp = new HashMap<>();

        for (int[] g : guards) {
            q.add(g);
            mp.put(g[0] + "," + g[1], 1);
            vis[g[0]][g[1]] = 1;
        }

        for (int[] w : walls) {
            mp.put(w[0] + "," + w[1], 1);
            vis[w[0]][w[1]] = 1;
        }

        for (int[] g : guards) {
            int r = g[0], c = g[1];
            dfs(r, c + 1, "r", vis, mp);
            dfs(r, c - 1, "l", vis, mp);
            dfs(r + 1, c, "d", vis, mp);
            dfs(r - 1, c, "u", vis, mp);
        }

        int cnt = 0;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                if (vis[i][j] == 0) cnt++;
        return cnt;
    }
}